src/inverted_index.c

```diff
@@ -968,6 +968,32 @@ size_t IR_NumEstimated(void *ctx) {
 
 #define FIELD_MASK_BIT_COUNT (sizeof(t_fieldMask) * 8)
 
+// Used to detemrmine if the field mask for the given doc id are valid based on their ttl:
+// ir->filterCtx.predicate
+static inline bool VerifyFieldMaskExpirationForDocId(IndexReader *ir, t_docId docId) {
+  // If there isn't a ttl information then the doc fields are valid
+  if (!ir->sctx || !ir->sctx->spec || !DocTable_HasExpiration(&ir->sctx->spec->docs, docId)) {
+    return true;
+  }
+
+  // doc has expiration information, create a field id array to check for expiration predicate
+  size_t numFieldIndices = 0;
+  // Use a stack allocated array for the field indices, if the field mask is not a single field
+  t_fieldIndex fieldIndicesArray[FIELD_MASK_BIT_COUNT];
+  t_fieldIndex* sortedFieldIndices = fieldIndicesArray;
+  if (ir->filterCtx.field.isFieldMask) {
+    numFieldIndices = IndexSpec_TranslateMaskToFieldIndices(ir->sctx->spec,
+                                                            ir->filterCtx.field.value.mask,
+                                                            fieldIndicesArray);
+  } else if (ir->filterCtx.field.value.index != RS_INVALID_FIELD_INDEX) {
+    sortedFieldIndices = &ir->filterCtx.field.value.index;
+    ++numFieldIndices;
+  }
```

Comment:
Why do we need to perform this on every read? it seems that all this info is constant in the context of the running query. We should do it once during the iterator creation.

Reply:
Expiration for fields is set per document.
You can only check if the fields expired when you reach the specific document.
Not sure I understand what you are suggesting.

Reply:
all the marked lines are not related to the current document and should be done once at the iterator constructor.